url0307c0e995_js_labs_js_collider_tgz-pJ8-org_jsl_tests_shmem_throughput_MainJ8/src/org/jsl/collider/AcceptorImpl.java

```diff
@@ -0,0 +1,685 @@
+
+                m_collider.executeInSelectorThread( new SessionStarter1(socketChannel) );
+            }
+
+            m_collider.executeInSelectorThread( m_starter3 );
+        }
+    }
+
+    private class SessionStarter1 extends ColliderImpl.SelectorThreadRunnable
+    {
+        private final SocketChannel m_socketChannel;
+
+        public SessionStarter1( SocketChannel socketChannel )
+        {
+            m_socketChannel = socketChannel;
+        }
+
+        public int runInSelectorThread()
+        {
+            try
+            {
+                final SelectionKey selectionKey = m_socketChannel.register( m_selector, 0, null );
+                m_collider.executeInThreadPool( new SessionStarter2(m_socketChannel, selectionKey) );
+            }
+            catch (final IOException ex)
+            {
+                /* Not necessary a framework problem,
+                 * can happen in a case if peer closed connection.
+                 */
+                if (s_logger.isLoggable(Level.FINE))
+                    s_logger.fine( m_localAddr + ": " + ex );
+                releaseMonitor();
+            }
+            return 0;
+        }
+    }
+
+    private class SessionStarter2 extends ThreadPool.Runnable
+    {
+        private final SocketChannel m_socketChannel;
+        private final SelectionKey m_selectionKey;
+
+        public SessionStarter2( SocketChannel socketChannel, SelectionKey selectionKey )
+        {
+            m_socketChannel = socketChannel;
+            m_selectionKey = selectionKey;
+        }
+
+        public void runInThreadPool()
+        {
+            startSession( m_socketChannel, m_selectionKey );
+        }
+    }
+
+    private class Starter1 extends ColliderImpl.SelectorThreadRunnable
+    {
+        public int runInSelectorThread()
+        {
+            m_lock.lock();
+            try
+            {
+                assert( m_state == STARTING_0 );
+                if (m_stop)
+                    return 0;
+                m_state = STARTING_1;
+                m_pendingOps = 1;
+            }
+            finally
+            {
+                m_lock.unlock();
+            }
+
+            try
+            {
+                m_selectionKey = m_serverChannel.register( m_selector, 0, AcceptorImpl.this );
+                m_collider.executeInThreadPool( new Starter2() );
+                return 0;
+            }
+            catch (final IOException ex)
+            {
+                /* Any exception here means a bug in Collider framework. */
+                if (s_logger.isLoggable(Level.WARNING))
+                    s_logger.warning( m_localAddr + ": " + ex + "." );
+            }
+
+            try
+            {
+                m_serverChannel.close();
+            }
+            catch (final IOException ex1)
+            {
+                if (s_logger.isLoggable(Level.WARNING))
+                    s_logger.warning( m_localAddr + ": " + ex1 + "." );
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 199:207, 357:365, 600:608, 631:639.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.